lib/constable/services/announcement_creator.ex

```diff
@@ -13,17 +13,29 @@ defmodule Constable.Services.AnnouncementCreator do
 
   def create(params, interest_names) do
     changeset = Announcement.create_changeset(%Announcement{}, params)
-    case Repo.insert(changeset) do
-      {:ok, announcement} ->
-        announcement
-        |> add_interests(interest_names)
-        |> Repo.preload(:user)
-        |> subscribe_author
-        |> email_and_subscribe_users
-        |> SlackHook.new_announcement
-
-        {:ok, announcement}
-      error -> error
+
+    case interest_names do
+      [] ->
+        {:error,
+         changeset
+         |> Ecto.Changeset.add_error(:interests, "you must select at least one interest")
+         |> Map.put(:action, :update)}
+
+      _ ->
+        case Repo.insert(changeset) do
+          {:ok, announcement} ->
+            announcement
+            |> add_interests(interest_names)
+            |> Repo.preload(:user)
+            |> subscribe_author
+            |> email_and_subscribe_users
+            |> SlackHook.new_announcement()
+
+            {:ok, announcement}
+
+          error ->
+            error
+        end
```

Comment:
I am having some kind of reaction to the nested `case` statement, but I cant really describe it. :thinking: I actually dont see anything wrong with it, but its triggering some kind of fear of future complexity if this direction continues. 

Im also not super experienced with the general culture of Elixir, and am wondering, are nested `case` statements typical? I did a web search and haphazardly clicked on this [SO qustion](https://stackoverflow.com/questions/50827551/elixir-how-to-avoid-deeply-nested-case-statements), which introduced me to the [`with` function](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#with/1):

```elixir
opts = %{width: 10}
with {:ok, width} <- Map.fetch(opts, :width),
     {:ok, height} <- Map.fetch(opts, :height) do
  {:ok, width * height}
else
  :error -> {:error, :wrong_data} # fail case reached, since no `opts[:height]`
end
```

I suppose something like that :point_up_2: would be how to deal with future complexity, if someone feels included to add more `case` nesting. 

Im not suggesting a change now, as I cant tell if its just adding unecessary complexity, and I do think this code is fine as-is. But I can share at least that I learned a new pattern from this :grin:, and am wondering if anyone else has any thoughts on nested `case` statements in elixir :alembic:

Reply:
@stratigos I didn't know about the `with` control. cool!